### ISSUE TITLE
Feature request?: PowerToys FancyZones and/or Windows 11 22H2 Snap layouts support

### ISSUE THREAD

**Issue by streger** [RamonUnch/AltSnap#308]
Hello,

First of all a BIG <3 thanks for this awesome application and your continued support for it.
I'm a very big fan of this application and have used it since the early days when it was produced by Stefan.
I've tried my best to check if there's already an integrated support for PowerToys FancyZones and/or Windows 11 22H2 Snap layouts but as much as I can see there isn't one currently.
I'm using the AltSnap just because of one (main) feature, the Alt + mouse left click to drag windows (sometimes).
I'm also a fan of PowerToys FancyZones and have already zones setup there, which works flawlessly via title-bar drag. That's where I'm saddened when I Alt+mouse drag a window but it doesn't snap to a FancyZone.
I'm also considering using just the latest Windows 11 22H2 snap layouts feature, since it has somewhat reasonable keyboard support finally and maybe that would be an option to integrate here as well. If I Alt+mouse drag to the screen top for Windows 11 layouts snap to appear, it doesn't.

I know that AltSnap also has its own zones setup, but for me that seems like an overburn (a third zones system on the same OS).

Is there already a support for any of two mentioned somehow, or I'm just a slow learner?

Please share some insight,

Thanks

**Comment by RamonUnch** [comment]
The issue there is that it is impossible for AltSnap to properly integrate with FancyZones because FancyZones handles its own zones entirely himself, there is no way for any other program to ask to FancyZones to snap a window somewhere. FancyZones is a black box with no interface.

The same problem occurs with native window snapping. Even Aero-Snap has no programmatic interface. So when you snap a window via AltSnap it cannot be the same than with normal snapping because Microsoft does not offer any API to do so. And this is the same for Windows 11 Snap Layouts.

I actually would never have spent countless hours in programming a snap layout system if I could have avoided it. In the same manner Stefan could never have spent hours in programming an aero-snapping equivalent if it was not needed.

The things that are possible to improve FancyZones integration are:
1) Use snap layout and set it up as your FancyZones layout (this is easily done with test windows, that you can snap with FancyZones).
2) Check the Restore FancyZones window option in the advanced section of the .ini file.

It is possible for AltSnap to see that a window was snapped by FancyZones because it adds a flag to the window.
It is also possible (through an [undocumented hack](https://github.com/RamonUnch/AltSnap/blob/2869533c126be0f2ff1c16f94fd28e954e436607/unfuck.h#L961-L976) that I came up with) to detect that a window was snapped by Windows aero-snapping or onto a Windows 11 zone. So They can be restored when you Alt+Drag a natively snapped window.

So in conclusion If you want Alt+drag to properly interfaced with FancyZones the best way would be the [discontinued Power Drag](https://github.com/microsoft/PowerToys/pull/6979#issuecomment-733266923). Maybe someone will continue the branch, but I prefer to stay an independent project and not be a part of PowerToys.

Many of my users actually stopped using FancyZones and just use AltSnap to handle snap layout. If you set `Move` as a left click title-bar action, you will also get snapping when you move the window via their titlebar.
Personally I almost never click a titlebar anymore except to test AltSnap.

**Comment by streger** [comment]
Thank you very much for your most elaborative answer possible. It was very informative. Unfortunately, it's was also what I thought was the answer from reading through other issues here. I knew Microsoft is to blame, I just cannot believe that still in 2022 they still are not kind to world of Windows customization, at least enough to provide API's. And, just to think of the world of possibilities there, even for them, probably also some lucrative.
On the other note, I also use [MPC-HC](https://github.com/clsid2/mpc-hc) for video playing and it has this feature to just drag the whole video window with one click/drag by clicking anywhere in the window screen. That also works with both FancyZones and Windows snap layouts. But I think that's a completely different approach then what AltSnap is using.
It seems that I will in the near future anyhow stop using one of those other two, and finally setup myself the AltSnap snap layouts.
Thanks again, and please manage this ticket by closing it. People who are interested in the same will find it in the archive anyhow.

**Comment by RamonUnch** [comment]
>On the other note, I also use [MPC-HC](https://github.com/clsid2/mpc-hc) for video playing and it has this feature to just drag the whole video window with one click/drag by clicking anywhere in the window screen. That also works with both FancyZones and Windows snap layouts. But I think that's a completely different approach then what AltSnap is using.

It is indeed easy for an application to handle movement when clicking anywhere as the window can identify any parts of it area like a titlebar. To get more info this is the WM_NCHITTEST message, you can see the result using the **identify window** option in the blacklist tab. Titlebar responds 2, windows will move the window in the case you such such area.
